### PR TITLE
GFSSF Relevance Validator

### PR DESCRIPTION
**Pull request by kelvin-lu**
Just to have a PR up for the time being + see file diffs. test cases have not been written, and there are bound to be bugs, so not ready4review.

### CODE REVIEW

**Comment by kelvin-lu** [comment]
Apologies for the force push, I accidentally rebased and then also merged and created a diff that was the size of our entire commit history from the past month.

File: tests/end_to_end/test_validation_end_to_end.py [hunk 1]
```diff
@@ -90,15 +91,15 @@ def _import(modname):
...
+            q = 1
             X, y, coef = make_regression(
                 n_samples=50, n_features=p, n_informative=q, coef=True,
                 shuffle=True, random_state=1)
 
             # informative columns are 'A', 'B'
             # uninformative columns are 'Z_0', ..., 'Z_11'
             columns = []
-            informative = list('AB')
+            informative = list('DCBA')
```
Comment (micahjsmith) [inline tests/end_to_end/test_validation_end_to_end.py]: something wrong here, we want len(informative) == q

File: tests/end_to_end/test_validation_end_to_end.py [hunk 2]
```diff
@@ -185,29 +186,26 @@ def call_validate_all(pr=None):
...
+    new_feature_str = make_feature_str('A')
     username = 'charlie'
-    featurename = 'B'
+    featurename = 'A'
     submit_feature(repo, contrib_dir, username, featurename, new_feature_str)
-    call_validate_all(pr=3)
+    with pytest.raises(CalledProcessError):
+        call_validate_all(pr=3)
 
-    # merge PR with master
```
Comment (micahjsmith) [inline tests/end_to_end/test_validation_end_to_end.py]: why remove this part of the test?

File: tests/validation/test_gfssf_validator.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        self.assertEqual(0, same_val_h, 'Expected all ones to be zero entropy')
+
+        same_val_arr_zero = np.zeros((50, 1))
+        same_multi_val_arr = np.concatenate(
+            (same_val_arr_ones, same_val_arr_zero), axis=1)
+        same_multi_val_h = gfv._calculate_disc_entropy(same_multi_val_arr)
+        self.assertEqual(
+            0,
+            same_multi_val_h,
+            msg='Expected all ones to be zero entropy')
```
Comment (micahjsmith) [inline tests/validation/test_gfssf_validator.py]: not all ones anymore

File: tests/validation/test_gfssf_validator.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+        diff_val_arr = np.concatenate((same_val_arr_ones, same_val_arr_zero))
```
Comment (micahjsmith) [inline tests/validation/test_gfssf_validator.py]: add axis=0 explicitly to help deter confusion w/r/t previous case

File: tests/validation/test_gfssf_validator.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+import pandas as pd
+
+import ballet.validation.gfssf_validator as gfv
+
+
+class GFSSFValidadatorTest(unittest.TestCase):
+    def setUp(self):
+        pass
+
+    def test_discrete_entropy(self):
```
Comment (micahjsmith) [inline tests/validation/test_gfssf_validator.py]: split these into separate test cases (unit testing a lot cleaner than the end-to-end testing)

i.e. `test_discrete_entropy_same_val`

File: tests/validation/test_gfssf_validator.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        same_val_arr_ones = np.ones((50, 1))
+        same_val_h = gfv._calculate_disc_entropy(same_val_arr_ones)
+        self.assertEqual(0, same_val_h, 'Expected all ones to be zero entropy')
+
+        same_val_arr_zero = np.zeros((50, 1))
```
Comment (micahjsmith) [inline tests/validation/test_gfssf_validator.py]: `def test_discrete_entropy_multi_col_same_val`

File: tests/end_to_end/test_validation_end_to_end.py [hunk 7]
```diff
@@ -1,9 +1,10 @@
 import os
-from subprocess import check_call
+import tempfile
+from subprocess import CalledProcessError, check_call
 from textwrap import dedent
 from types import ModuleType
 from unittest.mock import patch
-
+import pytest
```
Comment (micahjsmith) [inline tests/end_to_end/test_validation_end_to_end.py]: sort properly

File: tests/end_to_end/test_validation_end_to_end.py (hunk 2 above)
Comment (kelvin-lu) [reply tests/end_to_end/test_validation_end_to_end.py]: I changed this from an accepted feature to a redundant feature that should not be accepted. I can add back the original feature B as well.

Comment (micahjsmith) [reply tests/end_to_end/test_validation_end_to_end.py]: ah, no i get it

File: ballet/validation/gfssf_validator.py [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+        return np.zeros((n_samples, 1))
+    return asarray2d(np.concatenate(filtered_dfs, axis=1))
+
+
+class GFSSFAcceptanceEvaluator(FeatureAcceptanceEvaluator):
+    def __init__(self, X_df, y, features, lmbda_1=0, lmbda_2=0):
+        super().__init__(X_df, y, features)
+        self.y = asarray2d(y)
+        if (lmbda_1 <= 0):
+            lmbda_1 = _estimate_entropy(self.y) / 32
```
Comment (micahjsmith) [inline ballet/validation/gfssf_validator.py]: make 32 a class-level constant

File: ballet/validation/gfssf_validator.py [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+def _concat_datasets(dfs_by_src, n_samples, omit=None):
+    filtered_dfs = [np.array(dfs_by_src[x])
+                    for x in dfs_by_src if x is not omit]
+    if len(filtered_dfs) == 0:
+        return np.zeros((n_samples, 1))
+    return asarray2d(np.concatenate(filtered_dfs, axis=1))
+
+
+class GFSSFAcceptanceEvaluator(FeatureAcceptanceEvaluator):
+    def __init__(self, X_df, y, features, lmbda_1=0, lmbda_2=0):
```
Comment (micahjsmith) [inline ballet/validation/gfssf_validator.py]: default to `0.`

File: ballet/validation/gfssf_validator.py [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+        self.y = asarray2d(y)
+        if (lmbda_1 <= 0):
+            lmbda_1 = _estimate_entropy(self.y) / 32
+        if (lmbda_2 <= 0):
+            lmbda_2 = _estimate_entropy(self.y) / 32
+        self.lmbda_1 = lmbda_1
+        self.lmbda_2 = lmbda_2
+
+    def judge(self, feature):
+        feature_df = make_mapper([feature]).fit_transform(self.X_df, self.y)
```
Comment (micahjsmith) [inline ballet/validation/gfssf_validator.py]: instead use `feature.as_dataframe_mapper()`

File: ballet/validation/gfssf_validator.py [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+        n_samples, n_feature_clms = feature_df.shape
```
Comment (micahjsmith) [inline ballet/validation/gfssf_validator.py]: `n_feature_cols`

File: ballet/validation/gfssf_validator.py [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+        n_feature_clms_arr = n_feature_clms
+        n_feature_grps_arr = 1
+        feature_dfs_by_src = {}
+        for accepted_feature in self.features:
+            accepted_df = make_mapper(
```
Comment (micahjsmith) [inline ballet/validation/gfssf_validator.py]: `feature.as_dataframe_mapper`

File: ballet/validation/gfssf_validator.py [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+            logger.debug(
+                'Testing with omitted feature: {}'.format(
+                    omit or 'None'))
+            z = _concat_datasets(feature_dfs_by_src, n_samples, omit)
+            cmi = _estimate_conditional_information(feature_df, self.y, z)
+            logger.debug(
+                'Conditional Mutual Information Score: {}'.format(cmi))
+            cmi_omit = 0
+            n_clms_omit = 0
+            if omit is not '':
```
Comment (micahjsmith) [inline ballet/validation/gfssf_validator.py]: equivalent to `if not omit` if `omit` is strings

File: ballet/validation/gfssf_validator.py [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+                omit_df = feature_dfs_by_src[omit]
+                cmi_omit = _estimate_conditional_information(
+                    omit_df, self.y, z)
+                _, n_clms_omit = omit_df.shape
```
Comment (micahjsmith) [inline ballet/validation/gfssf_validator.py]: `n_cols_omit`

File: ballet/validation/gfssf_validator.py [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+                logger.debug('Omitted CMI Score: {}'.format(cmi_omit))
+            statistic = cmi - cmi_omit
+            threshold = lmbda_1 + \
```
Comment (micahjsmith) [inline ballet/validation/gfssf_validator.py]: separate into `def _compute_threshold`

File: ballet/validation/gfssf_validator.py (hunk 8 above)
Comment (kelvin-lu) [reply ballet/validation/gfssf_validator.py]: Currently taking suggestions for the name of this constant (little stumped)

Comment (micahjsmith) [reply ballet/validation/gfssf_validator.py]: `LAMBDA_1_ADJUSTMENT`
`LAMBDA_1_MAGIC`
`LAMBDA_1_FUDGE_FACTOR`
`LAMBDA_1_SCALE`

File: ballet/validation/entropy.py [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+from scipy.special import digamma
+from sklearn.neighbors import NearestNeighbors
+
+from ballet.util import asarray2d
+
+NUM_NEIGHBORS = 3  # Used in the sklearn mutual information function
+
+
+def calculate_disc_entropy(X):
+    # An exact calculation of the dataset entropy, using empirical probability
```
Comment (micahjsmith) [inline ballet/validation/entropy.py]: add docstring

File: ballet/validation/entropy.py [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+    n_samples, _ = X.shape
+    _, counts = np.unique(X, axis=0, return_counts=True)
+    empirical_p = counts * 1.0 / n_samples
+    log_p = np.log(empirical_p)
+    entropy = -np.sum(np.multiply(empirical_p, log_p))
+    return entropy
+
+
+def estimate_cont_entropy(X, epsilon=None):
+    """
```
Comment (micahjsmith) [inline ballet/validation/entropy.py]: update docstring, for example:
```
"""Estimate continuous entropy

Uses the Kraskov estimator [1] etc. etc.

Args:
    X (array-like): whatever x is

References:
    [1] <https:// ...>
"""

File: ballet/validation/entropy.py [hunk 18]
```diff
@@ -0,0 +1,165 @@
...
+    nn = NearestNeighbors(
+        metric='chebyshev',
+        n_neighbors=NUM_NEIGHBORS,
+        algorithm='kd_tree')
+    nn.fit(X)
+    if epsilon is None:
+        # If epsilon is not provided, revert to the Kozachenko Estimator
+        n_neighbors = NUM_NEIGHBORS
+        radius = None
+        # While we have non-zero radii, calculate for a larger k
```
Comment (micahjsmith) [inline ballet/validation/entropy.py]: zero

File: ballet/validation/entropy.py [hunk 19]
```diff
@@ -0,0 +1,165 @@
...
+    else:
+        ind = nn.radius_neighbors(
+            radius=epsilon.ravel(),
+            return_distance=False)
+        nx = np.array([i.size for i in ind])
+        return - np.mean(digamma(nx + 1)) + digamma(n_samples)
+
+
+def _is_column_discrete(col):
+    # Stand-in method to figure out if column is discrete
```
Comment (micahjsmith) [inline ballet/validation/entropy.py]: add the heuristics for integer values and integer-adjacent floating point values

File: ballet/validation/entropy.py [hunk 20]
```diff
@@ -0,0 +1,165 @@
...
+    uniques = np.unique(col)
+    return (uniques.size / col.size) < 0.05
+
+
+def _get_discrete_columns(X):
+    return np.apply_along_axis(_is_column_discrete, 0, X)
+
+
+def estimate_entropy(X, epsilon=None):
+    """
```
Comment (micahjsmith) [inline ballet/validation/entropy.py]: update docstring

File: ballet/validation/entropy.py [hunk 21]
```diff
@@ -0,0 +1,165 @@
...
+    cont_mask = ~disc_mask
+    # If our dataset is fully discrete/continuous, do something easier
+    if np.all(disc_mask):
+        return calculate_disc_entropy(X)
+    elif np.all(cont_mask):
+        return estimate_cont_entropy(X, epsilon)
+
+    # Separate the dataset into discrete and continuous datasets d,c
+    # We use the following reworking of entropy for efficient calculation:
+    # H(c,d) = sum_{x in d} p(x) * H(c(x)) + H(d)
```
Comment (micahjsmith) [inline ballet/validation/entropy.py]: $H(c,d) = sum_{x \in d} p(x) \times H(c(x)) + H(d)$

File: ballet/validation/entropy.py [hunk 22]
```diff
@@ -0,0 +1,165 @@
...
+            selected_cont_samples, selected_epsilon)
+        entropy += empirical_p[i] * conditional_cont_entropy
+    entropy += calculate_disc_entropy(disc_features)
+    if epsilon is None:
+        entropy = max(0, entropy)
+    return entropy
+
+
+def _calculate_epsilon(X):
+    """
```
Comment (micahjsmith) [inline ballet/validation/entropy.py]: update docstring

File: ballet/validation/entropy.py [hunk 23]
```diff
@@ -0,0 +1,165 @@
...
+    cont_features = X[:, ~disc_mask]
+    nn = NearestNeighbors(metric='chebyshev', n_neighbors=NUM_NEIGHBORS)
+    nn.fit(cont_features)
+    distances, _ = nn.kneighbors()
+    epsilon = np.nextafter(distances[:, -1], 0)
+    return asarray2d(epsilon)
+
+
+def estimate_conditional_information(x, y, z):
+    """
```
Comment (micahjsmith) [inline ballet/validation/entropy.py]: update docstring

File: ballet/validation/entropy.py [hunk 24]
```diff
@@ -0,0 +1,165 @@
...
+    epsilon = _calculate_epsilon(xyz)
+    h_xz = estimate_entropy(xz, epsilon)
+    h_yz = estimate_entropy(yz, epsilon)
+    h_xyz = estimate_entropy(xyz, epsilon)
+    h_z = estimate_entropy(z, epsilon)
+    return max(0, h_xz + h_yz - h_xyz - h_z)
+
+
+def estimate_mutual_information(x, y):
+    xy = np.concatenate((x, y), axis=1)
```
Comment (micahjsmith) [inline ballet/validation/entropy.py]: add docstring
